Pileup on Spark

tomwhite opened the pull request:
This addresses #1558 by using an implementation of #1988.

Running on a 6.6GB BAM file:
- PileupSpark with a shuffle took 4.35 minutes on a 6 node cluster
- PileupSpark with no shuffle (using the overlaps partitioner) took 2.96 minutes (30% faster)
- Pileup took 39.2 minutes (on a single node)

@droazen, please take a look and let me know what you think.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+
+    private static class ShufflePileupFunction implements FlatMapFunction<Tuple2<SimpleInterval, Iterable<GATKRead>>, AlignmentContext> {
```
I think that this could be useful outside the `PileupSpark` tool to create `JavaRDD<AlignmentContext>` easily. I'm working with pileups and this will be very useful for me...

tomwhite replied:
We could expose a method that returns an `JavaRDD<AlignmentContext>`. Also, do you need the reference bases for each context too? That needs a bit more work.

magicDGS replied:
Thanks a lot for the answer @tomwhite!

Actually it would be nice to have a way of using an implemented engine for a `LocusWalker` work with Spark. Thus, I think that this requires to have the reference bases for each context. But I don't know if this is out of the scope of this PR or if it is doable here. I would appreciate a lot if it could be done :)

tomwhite commented:
I've made some improvements to this PR, including:

* Made it easier to use the `joinOverlapping` method by making the function you supply only have to worry about one interval (shard) at a time. This simplifies the callers code, so PileupSpark (for example) is now shorter.
* Added some documentation.

I've also used the same technique to improve `AddContextDataToReadSpark` so that references are filled in on a per shard basis, rather than per read. In tests on a 6.6GB file I managed to get BaseRecalibratorSpark's runtime down from 10.61 minutes to 3.73 minutes, which is over 60% faster.

tomwhite commented:
I've improved PileupSpark so that it more closely mirrors the walker version. In particular, I've created a `LocusWalkerSpark` class that has a (shuffleless) method to obtain a

`JavaRDD<Tuple3<AlignmentContext, ReferenceContext, FeatureContext>>`

to mirror the

`apply(AlignmentContext, ReferenceContext, FeatureContext)`

method on `LocusWalker`.

It should be possible to use a similar approach to make Spark equivalents of other walkers, which will make porting the walkers to Spark easier.

What do you think @magicDGS, @droazen?

magicDGS commented:
Thanks for this, @tomwhite! Actually this is awesome for me, because I don't know too much about the Spark framework and it makes more easy the implementation of tools compatible with it.

In addition, I think that it will be awesome to have this for other type of walkers. Thanks a lot again!

tomwhite commented:
BTW this approach works for other walkers as well. I've converted ReadWalker and AssemblyRegionWalker to their Spark versions in this branch (which I'll submit after this goes in): https://github.com/broadinstitute/gatk/compare/tw_spark_walkers.

hammer commented:
cc @ryan-williams who has done a tremendous amount of work on handling skew when building pileups for https://github.com/hammerlab/guacamole

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/spark/SparkUtilsShardUnitTest.java:
```diff
@@ -0,0 +1,355 @@
+        JavaPairRDD<Locatable, Integer> readsPerInterval =
+                SparkUtils.shard(ctx, reads, TestRead.class, sequenceDictionary, shardBoundaries, false)
+                .flatMapToPair(new CountOverlappingReadsFunction());
+        assertEquals(readsPerInterval.collectAsMap(), ImmutableMap.of(intervals.get(0), 1, intervals.get(1), 7, intervals.get(2), 4));
```
Declare the Map of expected values separately above, to make it clearer here that you're asserting equality against the same values in both the shuffle and non-shuffle cases. Also applies to the tests below.

tomwhite replied:
Done

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/spark/SparkUtilsShardUnitTest.java:
```diff
@@ -0,0 +1,355 @@
+        assertFalse(it.hasNext());
+    }
+
+    static class TestRead implements Locatable {
```
Make private

droazen commented on src/test/java/org/broadinstitute/hellbender/utils/spark/SparkUtilsShardUnitTest.java:
```diff
@@ -0,0 +1,355 @@
+        assertEquals(next._2(), ImmutableList.of(reads.get(7), reads.get(8), reads.get(9), reads.get(10)));
+
+        assertFalse(it.hasNext());
+    }
```
Existing test cases are good, but I think we might also want a slightly larger-scale test that uses more than just a small handful of reads.

tomwhite replied:
The logic is heavily exercised by the integration tests for the various tools that use this feature (there are more to come after this PR), so the small-scale unit tests that cover certain edge cases, plus the integration tests, should be sufficient.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSpark.java:
```diff
@@ -45,4 +63,53 @@
     }
+
+    private static final int SHARD_SIZE = 10000;
+    private static final int SHARD_PADDING = 1000;
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSpark.java:
```diff
@@ -45,4 +63,53 @@
+    private static final int SHARD_SIZE = 10000;
+    private static final int SHARD_PADDING = 1000;
+
+    public static JavaPairRDD<GATKRead, ReadContextData> add(
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSpark.java:
```diff
@@ -45,4 +63,53 @@
+        final List<SimpleInterval> intervals = IntervalUtils.getAllIntervalsForReference(sequenceDictionary);
+        // use unpadded shards (padding is only needed for reference bases)
+        final List<ShardBoundary> intervalShards = intervals.stream()
+                .flatMap(interval -> Shard.divideIntervalIntoShards(interval, SHARD_SIZE, SHARD_SIZE, 0, sequenceDictionary).stream())
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSpark.java:
```diff
@@ -45,4 +63,53 @@
+                    }
+                });
+                // only include reads that start in the shard
+                return () -> Iterators.filter(transform, r -> r._1().getStart() >= shard.getStart());
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java:
```diff
@@ -0,0 +1,138 @@
+package org.broadinstitute.hellbender.tools.spark.pipelines;
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java:
```diff
@@ -0,0 +1,138 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+@CommandLineProgramProperties(summary = "Print read alignments in Pileup-style format on Spark",
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java:
```diff
@@ -0,0 +1,138 @@
+    @Override
+    public boolean requiresReads() { return true; }
+
+    @Argument(doc="the output file path", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java:
```diff
@@ -0,0 +1,138 @@
+    public boolean requiresReads() { return true; }
+
+    @Argument(doc="the output file path", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    protected String outputFile;
```
Yes. Updated the doc.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java:
```diff
@@ -0,0 +1,138 @@
+                .and(ReadFilterLibrary.MAPPED)
+                .and(ReadFilterLibrary.NOT_DUPLICATE)
+                .and(ReadFilterLibrary.PASSES_VENDOR_QUALITY_CHECK)
+                .and(ReadFilterLibrary.PRIMARY_ALIGNMENT);
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/PileupSpark.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    protected void runTool(final JavaSparkContext ctx) {
+        getAlignments(ctx).map(pileupFunction(metadata, outputInsertLength, showVerbose)).saveAsTextFile(outputFile);
```
This would indeed be nice, but it means that the function would capture the class instance variables in its closure, which means either that unnecessary data is transferred, or it would fail if there are objects that are not serializable (even if they are not used by the function).

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,102 @@
+package org.broadinstitute.hellbender.utils.spark;
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,102 @@
+package org.broadinstitute.hellbender.utils.spark;
+
+import autovalue.shaded.com.google.common.common.base.Function;
+import autovalue.shaded.com.google.common.common.collect.Iterators;
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,102 @@
+    private static final long serialVersionUID = 1L;
+
+    private static final int SHARD_SIZE = 10000;
+    private static final int SHARD_PADDING = 1000;
```
Done. Something to look at during the performance runs.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,102 @@
+    private static final int SHARD_PADDING = 1000;
+
+    @Argument(doc = "whether to use the shuffle implementation or not", shortName = "shuffle", fullName = "shuffle", optional = true)
+    public boolean shuffle = false;
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,102 @@
+        SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary();
+        List<SimpleInterval> intervals = hasIntervals() ? getIntervals() : IntervalUtils.getAllIntervalsForReference(sequenceDictionary);
+        final List<ShardBoundary> intervalShards = intervals.stream()
+                .flatMap(interval -> Shard.divideIntervalIntoShards(interval, SHARD_SIZE, SHARD_SIZE, SHARD_PADDING, sequenceDictionary).stream())
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,102 @@
+        return shardedReads.flatMap(getAlignmentsFunction(bReferenceSource, bFeatureManager, sequenceDictionary, getHeaderForReads()));
+    }
+
+    private static FlatMapFunction<Shard<GATKRead>, Tuple3<AlignmentContext, ReferenceContext, FeatureContext>> getAlignmentsFunction(
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,102 @@
+            final Set<String> samples = header.getReadGroups().stream()
+                    .map(SAMReadGroupRecord::getSample)
+                    .collect(Collectors.toSet());
+            LocusIteratorByState libs = new LocusIteratorByState(readIterator, DownsamplingMethod.NONE, false, false, false, samples, header);
```
Done. Added a method to allow the downsampling method to be specified.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/CoalescedRDD.java:
```diff
@@ -0,0 +1,91 @@
+import scala.reflect.ClassTag;
+import scala.reflect.ClassTag$;
+
+class CoalescedRDD<T> extends RDD<T> {
```
This a Spark 2 backport. I've added a note.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/CoalescedRDDPartition.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+class CoalescedRDDPartition implements Partition {
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/PartitionCoalescer.java:
```diff
@@ -0,0 +1,8 @@
+import org.apache.spark.rdd.PartitionGroup;
+import org.apache.spark.rdd.RDD;
+
+interface PartitionCoalescer {
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/RangePartitionCoalescer.java:
```diff
@@ -0,0 +1,44 @@
+    @Override
+    public PartitionGroup[] coalesce(int maxPartitions, RDD<?> parent) {
+        if (maxPartitions != parent.getNumPartitions()) {
+            throw new IllegalArgumentException("Cannot use " + getClass().getSimpleName() +
```
It's a limitation of the PartitionCoalescer which we are duplicating from Spark. RangePartitionCoalescer is more restrictive, so we have a check to ensure it doesn't change the number of partitions.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+     * @param <L> the {@link Locatable} type
+     * @return an RDD of {@link Shard} of overlapping {@link Locatable} objects
+     */
+    public static <L extends Locatable> JavaRDD<Shard<L>> shard(JavaSparkContext ctx, JavaRDD<L> locatables, Class<L> locatableClass,
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+    }
+
+    /**
+     * Join an RDD of reads with a set of intervals, and apply a function to process the reads that overlap each interval.
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+     * @param <T> the return type of <code>f</code>
+     * @return
+     */
+    private static <R extends Locatable, I extends Locatable, T> JavaRDD<T> joinOverlapping(JavaSparkContext ctx, JavaRDD<R> reads, Class<R> readClass,
```
Done

tomwhite commented on src/test/java/org/broadinstitute/hellbender/tools/spark/PileupSparkIntegrationTest.java:
```diff
@@ -0,0 +1,107 @@
+            args.add("--shuffle");
+        }
+        this.runCommandLine(args.getArgsArray());
+        File expected = new File(TEST_DATA_DIR, "expectedSimplePileup.txt");
```
Yes

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/FeatureManager.java:
```diff
@@ -80,7 +80,7 @@
      * pool of FeatureDataSources
      */
-    private final CommandLineProgram toolInstance;
+    private transient final CommandLineProgram toolInstance;
```
Possibly. The tests work with this change (this includes tests that load features that aren't in this branch), so it's useful for some cases.

tomwhite replied:
I looked into this more, and it seems that the problem was that Spark was trying to serialize some log4j classes, which are not serializable, and even Kryo can't serialize them. I think there's a fix that makes the logger fields transient and then reinststes them via `readObject` after deserialization. I'll add this to the next PR.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+     * @param intervals the {@link ShardBoundary} objects to create shards for, must be coordinate sorted
+     * @param useShuffle whether to use a shuffle or not
+     * @param <L> the {@link Locatable} type
+     * @return an RDD of {@link Shard} of overlapping {@link Locatable} objects
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+    public static <L extends Locatable> JavaRDD<Shard<L>> shard(JavaSparkContext ctx, JavaRDD<L> locatables, Class<L> locatableClass,
+                                                                SAMSequenceDictionary sequenceDictionary, List<ShardBoundary> intervals, boolean useShuffle) {
+
+        class ShardBoundaryShard implements Shard<L> {
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+                    }
+                    R read = peekingReads.next();
+                    if (overlaps(currentShard, read)) {
+                        currentReads.add(read);
```
Yes, it might be. There should be a way to downsample - even skip over partitions that have the same reads in them - but this could be tackled as a follow-on piece of work.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
 
+    /**
+     * Create an RDD of {@link Shard} from an RDD of coordinate sorted {@link Locatable} <i>without using a shuffle</i>.
+     * Each shard contains the {@link Locatable} objects that overlap it.
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+                return locatables.iterator();
+            }
+        }
+        List<ShardBoundary> paddedIntervals = intervals.stream().map(sb -> new ShardBoundary(sb.getInterval(), sb.getPaddedInterval(), true)).collect(Collectors.toList());
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/spark/SparkUtils.java:
```diff
@@ -103,4 +123,358 @@ public static boolean pathExists(final JavaSparkContext ctx, final Path targetPa
+        PeekingIterator<R> peekingReads = Iterators.peekingIterator(reads);
+        PeekingIterator<I> peekingShards = Iterators.peekingIterator(shards);
+        Iterator<Tuple2<I, Iterable<R>>> iterator = new AbstractIterator<Tuple2<I, Iterable<R>>>() {
+            // keep track of current and next, since reads can overlap two shards
```
We're assuming that shards are larger than reads (and padding is too). Since shard and padding size is configurable, this shouldn't be a problem. I'll add a fail-fast check for this constraint, as we discussed.

tomwhite commented on src/test/java/org/broadinstitute/hellbender/utils/spark/SparkUtilsShardUnitTest.java:
```diff
@@ -0,0 +1,355 @@
+        //                       [-----------------------]
+        //
+        //                      1                   2
+        //    1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSpark.java:
```diff
@@ -45,4 +63,53 @@
+                            //In those cases, we'll just say that nothing overlaps the read
+                            overlappingVariants = Collections.emptyList();
+                        }
+                        return new Tuple2<>(r, new ReadContextData(referenceBases, overlappingVariants));
```
No, it's not used, but the padding will likely give enough context and is a more efficient way of doing this. Address as a follow on?

tomwhite commented:
@droazen thanks for the review. I have now addressed all of your feedback, with the main changes being

* Revert the change to ShardBoundary that uses the padded interval as its interval, and use an anonymous class instead.
* Add a new join strategy (OVERLAPS_PARTITIONER) so that running BQSR can be done using the old way still.
* Add a check for overly long read sizes. If exceeded the job will fail with an exception.
* Be more conservative about the partition end point by using the maximum read length, rather than just the length of the read that happens the start the next partition.
* Rather than making a field in FeatureManager transient, do a better job of reinstating the field that is not serializable (a logger).

droazen commented on src/main/java/org/broadinstitute/hellbender/utils/recalibration/RecalibrationArgumentCollection.java:
```diff
@@ -108,6 +108,12 @@
+
+    @Argument(fullName="readShardPadding", shortName="readShardPadding", doc = "Each read shard has this many bases of extra context on each side.", optional = true)
+    public int readShardPadding = 1000;
+
```
These should not be in `RecalibrationArgumentCollection`, since they are specific to Spark (and to one particular Spark sub-implementation). Can you either move to `BaseRecalibratorSpark`/`BQSRPipelineSpark` or, if you'd prefer to avoid repetition, to your own `OverlapsPartitionerArgumentCollection`/`SparkSharderArgumentCollection` included by BQSR and its various pipelines? 

Also adjust the argument docs to make it clear that they only apply when using the `OVERLAPS_PARTITIONER` join strategy.

tomwhite replied:
Done

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSparkUnitTest.java:
```diff
@@ -34,7 +36,9 @@
         List<Class<?>> classes = Arrays.asList(Read.class, SAMRecord.class);
-        JoinStrategy[] strategies = JoinStrategy.values();
+        EnumSet<JoinStrategy> joinStrategies = EnumSet.allOf(JoinStrategy.class);
+        joinStrategies.remove(JoinStrategy.OVERLAPS_PARTITIONER); // exclude overlaps partitioner, since it needs sequence dictionary not available for mock source
```
Can you extend the mock so that `ReferenceMultiSource.getReferenceSequenceDictionary()` returns a value, so that we can have test coverage of `JoinStrategy.OVERLAPS_PARTITIONER` in this class? Eg., 

        when(mockSource.getReferenceSequenceDictionary(null)).thenReturn(myDictionary);

tomwhite replied:
Done

droazen commented on src/test/java/org/broadinstitute/hellbender/engine/spark/SparkSharderUnitTest.java:
```diff
@@ -0,0 +1,393 @@
+        // Intervals
+        //     [-----]
+        //                 [---------]
+        //                       [-----------------------]
```
What happened to the partition read extents in this diagram (could have sworn they were here when you presented last week)?

tomwhite replied:
I was a bit over-zealous in re-writing that test. I've added the extents back to the diagram.

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/AddContextDataToReadSpark.java:
```diff
@@ -25,24 +42,87 @@
+            final JavaRDD<GATKRead> reads, final ReferenceMultiSource referenceSource,
+            final JavaRDD<GATKVariant> variants, final JoinStrategy joinStrategy,
+            final SAMSequenceDictionary sequenceDictionary,
+            final int shardSize, final int shardPadding) {
```
Document the new `shardSize` and `shardPadding` args, make it clear that they apply only when `JoinStrategy.OVERLAPS_PARTITIONER` is specified, and suggest values to be passed in for these args when other JoinStrategies are being used (0?).

tomwhite replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/LocusWalkerSpark.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Each read shard has this many bases of extra context on each side.
+     */
+    private static final int SHARD_PADDING = 1000;
```
Can you expose these as arguments in `LocusWalkerSpark`, like you did for BQSR?

tomwhite replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/RangePartitionCoalescer.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * A {@link PartitionCoalescer} that allows a range of partitions to be coalesced into groups.
+ */
+class RangePartitionCoalescer implements PartitionCoalescer, Serializable {
```
Simple unit test for this class? (unless it's a copy of code from Spark2)

tomwhite replied:
Done

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java:
```diff
@@ -0,0 +1,396 @@
+                    if (toRightOf(currentShard, peekingLocatables.peek(), sequenceDictionary)) {
+                        break;
+                    }
+                    R read = peekingLocatables.next();
```
Still some code in here and below that uses the term "read" -- should generalize (eg., `currentLocatable`)

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java:
```diff
@@ -0,0 +1,396 @@
+                    if (read.getContig() != null) {
+                        int size = read.getEnd() - read.getStart() + 1;
+                        if (size > maxLocatableLength) {
+                            throw new UserException(String.format("Max size of locatable exceeded. Max size is %s, but locatable size is %s. Try increasing size and/or padding. Locatable: %s", maxLocatableLength, size, read));
```
`size and/or padding` -> `shard size and/or padding`

droazen commented on src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java:
```diff
@@ -0,0 +1,396 @@
+                nextContigIndex = sequenceDictionary.getSequences().size();
+            }
+            if (intervalContigIndex == nextContigIndex) { // same contig
+                addPartitionReadExtent(extents, i, current.getContig(), current.getStart(), next.getStart() + maxLocatableLength);
```
So we're now overestimating the extent of each partition by `maxLocatableLength`, and this is set to 1000 by default, correct? Can you briefly explain the performance implications of this? I assume that we'll just potentially pull in a few extra partitions per interval that don't actually contain reads overlapping the interval?

tomwhite commented:
Thanks for the latest review comments @droazen. I've addressed them all. I'll merge this once the tests pass.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -44,7 +45,13 @@
+    private void readObject(java.io.ObjectInputStream in)
+            throws IOException, ClassNotFoundException {
+        in.defaultReadObject();
+        logger = LogManager.getLogger(this.getClass()); // Logger is not serializable (even by Kryo)
```
Yes, it looks like that would fix it. That change can be handled in #2176 if folks agree about moving to SLF4J (I would support it).

tomwhite replied:
@cmnbroad I'm not sure about the serialization cost (once per task, so maybe not a problem). The issue here is that `FeatureManager` has a `CommandLineProgram` field, and we want to broadcast `FeatureManager`. In the original PR I made the field transient, which seemed to work OK, but changed it after review feedback. We could revisit this if you think it's worth looking at.
